service/awsconfig/v7/templates/cloudformation/guest/lifecycle_hooks.go

```diff
@@ -0,0 +1,13 @@
+package guest
+
+const LifecycleHooks = `{{define "lifecycle_hooks"}}
+  NodeDrainerLifecycleHook:
+    Type: "AWS::AutoScaling::LifecycleHook"
+    Properties:
+      AutoScalingGroupName:
+        Ref: {{ .LifecycleHooks.NodeDrainer.Name }}
```

Comment:
This is wrong the ref should be `{{ .ASGType }}AutoScalingGroup`. That will render as `WorkerAutoScalingGroup` in the template and matches the CloudFormation resource name.

Getting the name via the EC2 API won't work because the stack hasn't executed yet so the resources don't exist.

Reply:
The ref is set to `WorkerAutoScalingGroup`, which is what it should be. I just do not want to hard code this in another place but get it from the key package. I want to refactor this at some point in the other templates as well but this is not my focus right now.

Reply:
I don't get why you're getting the ASG name in the current state?

You can just use the CloudFormation `Ref` function for that. We use the Adapter to get data from the EC2 API when its something external to the stack that we need.

Reply:
This should just be a template change. The lifecycle_hooks template just needs to reference the ASG created in the autoscaling_groups template. This simplifies the Go code a lot. Hope that helps.

Reply:
> I don't get why you're getting the ASG name in the current state?

I don't. The ASG name is hard coded in the key package. See https://github.com/giantswarm/aws-operator/pull/800/files#diff-a6dd1d9b43f4a81118882f7cb4912dd0R36. The adapter then sets the key package value for the template. See https://github.com/giantswarm/aws-operator/pull/800/files#diff-5875de7f0944b609a854823fdf0da710R19.

> This should just be a template change.

Not sure what you mean with that. My intention is to make the ASG name configurable in one place in the long run. I do not want to copy the ASG name from another template again.

Reply:
By template change I mean we use CloudFormation templating to link the AutoScalingGroup resource to the LifecycleHook resource.

The ASG resource is defined here and not in the key function.
https://github.com/giantswarm/aws-operator/blob/51e3d84066c3624b5dca6b1084093b9aa6f8b2cf/service/awsconfig/v7/templates/cloudformation/guest/auto_scaling_group.go#L4

In the lifecycle hooks template we should reference it by CF resource name.

```
    Properties:
      AutoScalingGroupName:
        Ref: {{ .ASGType }}AutoScalingGroup
```

The templating here is a bit confusing. When its rendered we need it to be.

`        Ref: workerAutoScalingGroup`

Using the CF templating this way is the right approach and means we don't need to use the Adapter package for this.

Reply:
I want to replace `{{ .ASGType }}AutoScalingGroup` with `{{ .LifecycleHooks.NodeDrainer.Name }}`. I want to replace this also for the auto scaling template in the long run to have it the same and to not have magic strings in different templates which will just fail when you change them in one place and forget the other. This is why I am doing it how I do it.

Reply:
I noticed that the naming of the structures here is wrong which probably leads to confusion while we talk. I will change this a bit to be correct.

Reply:
Cool thanks I think I get this and where the confusion came from.

Totally agree with replacing `{{ .ASGType }}AutoScalingGroup` in the template. We should be explicit and set the full autoscaling group name.

❤️ the new `.LifecycleHooks.ASG.Name` that is much clearer for me.